<!DOCTYPE html>
<html lang="en"><head>
<script src="lec03-multiple-genomics_files/libs/clipboard/clipboard.min.js"></script>
<script src="lec03-multiple-genomics_files/libs/quarto-html/tabby.min.js"></script>
<script src="lec03-multiple-genomics_files/libs/quarto-html/popper.min.js"></script>
<script src="lec03-multiple-genomics_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="lec03-multiple-genomics_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="lec03-multiple-genomics_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="lec03-multiple-genomics_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="lec03-multiple-genomics_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.433">

  <meta name="author" content="Lucy Gao">
  <meta name="dcterms.date" content="2024-01-23">
  <title>Multiple Testing and Genomics</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="lec03-multiple-genomics_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="lec03-multiple-genomics_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
  </style>
  <link rel="stylesheet" href="lec03-multiple-genomics_files/libs/revealjs/dist/theme/quarto.css">
  <link href="lec03-multiple-genomics_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="lec03-multiple-genomics_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="lec03-multiple-genomics_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="lec03-multiple-genomics_files/libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="lec03-multiple-genomics_files/libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="lec03-multiple-genomics_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Multiple Testing and Genomics</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Lucy Gao 
</div>
</div>
</div>

  <p class="date">January 23, 2024</p>
</section>
<section id="todays-science-genomics" class="slide level1">
<h1>Today’s science: genomics</h1>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="microarray.png"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="microarray-chip.png"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
<aside class="notes">
<ul>
<li>Expression level: counts mRNA transcripts<br>
</li>
<li>From “single gene” study (e.g.&nbsp;genetics) to lots of genes (genomics)</li>
<li>First microarrays, then NGS</li>
<li>Now also single-cell, spatial single-cell, etc.</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="differential-gene-expression" class="slide level1">
<h1>Differential gene expression</h1>
<p>Suppose we have the expression level measurements for <span class="math inline">\(M\)</span> genes on <span class="math inline">\(n\)</span> observations, where observations belong to one of two (labelled) experimental conditions or populations.</p>
<div class="fragment">
<p><br></p>
<p><strong>Questions</strong>: Across the two (labelled) conditions/populations/etc., is gene <span class="math inline">\(j\)</span> expressed at different levels, for all <span class="math inline">\(j = 1, 2, \ldots, M\)</span>?</p>
</div>
<div class="fragment">
<p><br></p>
<p><strong>Statistically</strong>, a (large-scale) two-group testing problem:</p>
<div>
<ul>
<li class="fragment">View observations in each group as sample from superpopulation</li>
<li class="fragment">Compare the marginal (per-gene) distributions for sampling from superpopulations (i.e.&nbsp;<span class="math inline">\(F_X\)</span> and <span class="math inline">\(F_Y\)</span>) with respect to a chosen functional, e.g.&nbsp;mean or (sort of) median</li>
</ul>
</div>
<aside class="notes">
<ul>
<li>Eg. Mutant vs wild type mice</li>
<li>Eg. Healthy vs cancerous tissue</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div>
</section>
<section id="scientific-goals" class="slide level1">
<h1>Scientific goals</h1>
<p>What is the role of these hypothesis tests?</p>
<div>
<ul>
<li class="fragment">In a nutshell, screening! Genes that “pass” through the filter are subjected to expensive and thorough validation experiments.</li>
</ul>
</div>
<div class="fragment">
<p>Overall philosophy:</p>
<div>
<ul>
<li class="fragment"><p>We want to get a moderately sized list of genes that seem like promising candidates for further investigation, and we want it fast and cheap.</p></li>
<li class="fragment"><p>The “cost” of a false positive is that we waste resources on validating a red herring.</p></li>
<li class="fragment"><p><strong>Key idea</strong>: We can live with (say) 10% of our resources going to waste in the follow-up phase.</p></li>
<li class="fragment"><p>After all, we have already saved a bunch of time and money by not investigating every gene in the full list!</p></li>
</ul>
</div>
<aside class="notes">
<ul>
<li>The genes aren’t carefully chosen. They’re just whatever we can measure. (measurement is cheap, since it’s parallel!)</li>
<li>So it makes sense to use the hypothesis test as a simple screen (rather than a super confirmational paradigm)</li>
<li>The rejections are not the end of a story. They are the beginning of many.</li>
<li>Validation: bioinformatic efforst, wet lab efforts</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div>
</section>
<section id="screen-and-clean-in-biology" class="slide level1">
<h1>Screen and clean in biology</h1>
<p>This “screen and clean” philosophy is pervasive in high-throughput biological data contexts.</p>
<p><br></p>
<div class="fragment">
<p>Ex. <strong>Genome-wide association studies</strong>: Measure tons of genetic markers and an observable characteristic. Test for association between all of the genetic variant and observable characteristics.</p>
</div>
<div class="fragment">
<p><br></p>
<p>Ex. <strong>Neuroimaging</strong>: Simultaneously measure activity across the whole brain; divide up into 3d pixels (voxels), and test for differences between condition at all voxels.</p>
<aside class="notes">
<ul>
<li>Realy this is a story about high throuput biology.</li>
<li>even more generally: whenever there is lots of data and there is complex and/or noisy data, you will probably see multiple testing …<br>
</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div>
</section>
<section id="what-do-we-care-about" class="slide level1">
<h1>What do we care about?</h1>
<p>Naively, based on last week, might say we care about the FWER: probability of falsely rejecting at least one null hypothesis. That was (sometimes) sensible for last week.</p>
<div class="fragment">
<p><br></p>
<p><strong>This is not sensible this week.</strong></p>
<div>
<ul>
<li class="fragment"><p>Why should we behave like we care so much about a single false positive? It’s not a catastrophe, it’s just inconvenient!</p></li>
<li class="fragment"><p>With thousands or millions of tests, FWER controlling procedures can return very few rejections indeed. Is such an aggressive screen useful?</p></li>
</ul>
</div>
<p><br></p>
</div>
<div class="fragment">
<p>A new concept emerges: we care about the <strong>proportion of resources wasted on red herrings in the validation experiments</strong>, as well as the <strong>number of validation experiments we get to do</strong>.</p>
<aside class="notes">
<ul>
<li>Think about Bonferroni: alpha/M</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div>
</section>
<section id="false-discovery-proportion" class="slide level1">
<h1>False discovery proportion</h1>
<p>On the <strong>proportion of resources wasted on red herrings in the validation experiments</strong>:</p>
<ul>
<li>Assume that the follow-up cost is the same for each gene</li>
<li>Assume that cost is wasted for null genes</li>
</ul>
<div class="fragment">
<p>Let <span class="math inline">\(X \sim F_X\)</span> represent the data with which we test <span class="math inline">\(H_{0j}: T_j(F_X) = 0\)</span> for all <span class="math inline">\(j = 1, 2, \ldots, p\)</span>. Let the follow-up cost for each gene be <span class="math inline">\(C\)</span>.</p>
</div>
<div class="fragment">
<p>Then, the “wasted” cost over the total cost is: <span class="math display">\[
\begin{align}
\text{FDP}(X) &amp;=\frac{C | \{j : F_X \in \mathcal{F}_{0j} \text{ and } p_j(X) \leq \alpha\}|}{C | \{ j: p_j(X) \leq \alpha \}|}, ~~~ \text{ for } \mathcal{F}_{0j} \equiv \{ F: T_j(F) = 0\} \\
&amp;= \frac{| \{j : F_X \in \mathcal{F}_{0j} \text{ and } p_j(X) \leq \alpha\}|}{| \{ j: p_j(X) \leq \alpha \}|}  \\
&amp;= \frac{\text{Number of false discoveries}}{\text{Number of discoveries}} ~~~~ \text{(nb. take convention of } 0/0 = 0 \text{.)}
\end{align}
\]</span></p>
<aside class="notes">
<ul>
<li>Numerator: Resources wasted on chasing red herrings</li>
<li>Denominator: Total follow-up efforts</li>
<li>Numerator: Total number of falsely rejected</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div>
</section>
<section id="fdp-via-contingency-table" class="slide level1">
<h1>FDP via contingency table</h1>
<p>Suppose that we test <span class="math inline">\(H_{0j}: T_j(F_X) \in \Theta_0\)</span> for all <span class="math inline">\(j = 1, 2, \ldots, M\)</span> using a random variable <span class="math inline">\(X \sim F_X\)</span>. Define the following notation:</p>
<table>
<thead>
<tr class="header">
<th></th>
<th><span class="math inline">\(H_{0j}\)</span> retained</th>
<th><span class="math inline">\(H_{0j}\)</span> rejected</th>
<th>Total</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><span class="math inline">\(H_{0j}\)</span> true</td>
<td><span class="math inline">\(TN(X)\)</span></td>
<td><span class="math inline">\(FD(X)\)</span></td>
<td><span class="math inline">\(M_0\)</span></td>
</tr>
<tr class="even">
<td><span class="math inline">\(H_{0j}\)</span> false</td>
<td><span class="math inline">\(FN(X)\)</span></td>
<td><span class="math inline">\(TD(X)\)</span></td>
<td><span class="math inline">\(M_1\)</span></td>
</tr>
<tr class="odd">
<td>Total</td>
<td><span class="math inline">\(N(X)\)</span></td>
<td><span class="math inline">\(R(X)\)</span></td>
<td><span class="math inline">\(M\)</span></td>
</tr>
</tbody>
</table>
<div class="fragment">
<p>Then, can write the false discovery proportion as:</p>
<p><span class="math display">\[ \text{FDP}(X) = \frac{\text{FD}(X)}{\max \{1, R(X)\}}.\]</span> This computes <em>vertically</em>. Contrast the <em>horizontal</em> false positive proportion (<span class="math inline">\(\frac{\text{FD}(X)}{M_0}\)</span>) or FWER.</p>
<aside class="notes">
<ul>
<li>Verbally go through what are random variables</li>
<li>Traditional point null hypothesis criteria: don’t care about second row when defining error rates</li>
<li>can think about this as vaguely bayesian flavoured</li>
<li>you will investigate FPP in homework</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div>
</section>
<section id="fdp-and-fdr" class="slide level1">
<h1>FDP and FDR</h1>
<p>We have a compelling reason to care about the FDP in this context:</p>
<p><span class="math display">\[ \text{FDP}(X) = \frac{\text{FD}(X)}{\max \{1, R(X)\}}.\]</span></p>
<div class="fragment">
<p>But we can’t “control the FDP” in any single study: it’s a random variable. Instead, think about controlling its expectation, the <em>false discovery rate</em>:</p>
<p><span class="math display">\[ \text{FDR} = \mathbb{E}_{F_X}[\text{FDP}(X)].\]</span></p>
</div>
<div class="fragment">
<p>More scientifically: if we were to <strong>repeat</strong> the process of a differential gene expression analysis study followed up by validation experiments on the discoveries, then the average proportion of resources wasted per study across these repeats is the FDR.</p>
<ul>
<li>Assuming again follow-up cost uniform across genes.</li>
</ul>
</div>
</section>
<section id="your-turn-interpreting-fdr" class="slide level1">
<h1>Your turn: interpreting FDR</h1>
<p>Imagine you are working with a genome scientist to analyze their RNA sequencing data. You apply a procedure that controls FDR at level <span class="math inline">\(10\%\)</span> to their data set, and reject 100 null hypotheses. Your collaborator says:</p>
<blockquote>
<p>“Oh, so no more than 10 genes out of that 100 really are differentially expressed?”</p>
</blockquote>
<p>What is your answer?</p>
<div class="fragment">
<p><br></p>
<p>What if your collaborator had said instead:</p>
<blockquote>
<p>“Oh, so about 10 genes out of that 100 really are differentially expressed?”</p>
</blockquote>
<p>What is your answer?</p>
</div>
</section>
<section id="follow-up-interpreting-fdr" class="slide level1">
<h1>Follow-up: interpreting FDR</h1>
<p><span class="math display">\[ \text{FDR} = \mathbb{E}_{F_X} \left [\frac{\text{FD}(X)}{\max \{1, R(X)\}} \right ]\]</span></p>
<div class="fragment">
<ol type="1">
<li>FDR = <span class="math inline">\(\alpha\)</span> clearly doesn’t imply that <span class="math inline">\(\text{R(X)} \cdot \alpha\)</span> upper bounds the number of false discoveries.</li>
</ol>
<div>
<ul>
<li class="fragment"><p>In fact, since <span class="math inline">\(R(X)\)</span> is random, this is a nonsensical thing to say.</p></li>
<li class="fragment"><p>More reasonable to want a probabilistic upper bound for <span class="math inline">\(R(X)\)</span>.</p></li>
</ul>
</div>
</div>
<div class="fragment">
<ol start="2" type="1">
<li>We know averages across realizations. But this is asking about a single realization. We could only be confident in relating the two if we knew about <em>concentration</em>.</li>
</ol>
<div>
<ul>
<li class="fragment"><p>How much of the distribution of FDP is concentrated around the mean?</p></li>
<li class="fragment"><p>The variance of FDP also gives us some information about concentration.</p></li>
</ul>
</div>
<aside class="notes">
<ul>
<li>Lots of literature about probabilistically bounding <span class="math inline">\(R(X)\)</span>; if you’re interested, you might be abke to pick an ISSI talk about it!</li>
<li>exceedance probabilitiy: find value gamma such that prob that <span class="math inline">\(R(X)\)</span> is less than gamma is controlled</li>
<li>Some papers about formulas for variance of FDP (e.g.&nbsp;art owen 2005); doesn’t seem to have caught on.</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div>
</section>
<section id="comparing-fdr-to-fwer" class="slide level1">
<h1>Comparing FDR to FWER</h1>
<p><span class="math display">\[ \text{FDR} = \mathbb{E}_{F_X} \left [\frac{\text{FD}(X)}{\max \{1, R(X)\}} \right ], ~~~~~ \text{FWER} = \mathbb{P}_{F_X} [ \text{FD}(X) \geq 1 ].\]</span></p>
<div>
<ul>
<li class="fragment"><p>Intuitively, the FDR is more adaptive to non-null:null ratio and number of tests</p></li>
<li class="fragment"><p>For FDR, adding discoveries (rejecting true non-nulls) increases the denominator, allowing us to increase the numerator; logic holds no matter how many tests</p></li>
<li class="fragment"><p>But for FWER, only ever allowed to reject one true null, no matter how many non-nulls and no matter how many tests</p></li>
<li class="fragment"><p>Formally, FWER is always larger than the FDR, with equality when all tested hypotheses are null (Homework)</p></li>
</ul>
</div>
</section>
<section id="your-turn-interpretation-redux" class="slide level1">
<h1>Your turn: interpretation redux</h1>
<p>Imagine you are working with a genome scientist to analyze their microarray data. You apply a procedure that controls FDR at level <span class="math inline">\(10\%\)</span> to their data set, and reject 100 null hypotheses. Your collaborator says:</p>
<blockquote>
<p>“I’ve decided to focus on just the 10 genes in that rejection list that have the smallest unadjusted p-values. What can we say about those 10 genes?”</p>
</blockquote>
<p>What is your answer?</p>
<div class="fragment">
<p>What if instead they said:</p>
<blockquote>
<p>“I’ve decided to focus on just the 10 genes in that rejection list that are in a particular biological pathway. What can we say about those 10 genes?”</p>
</blockquote>
<p>What is your answer?</p>
</div>
</section>
<section id="no-subset-property-fdr" class="slide level1">
<h1>No subset property, FDR</h1>
<p>We know that the following quantity is below 0.1: <span class="math display">\[ \text{FDR} = \mathbb{E}_{F_X} \left [\frac{\text{FD}(X)}{\max \{1, R(X)\}} \right ].\]</span></p>
<div class="fragment">
<p>Your collaborator is tempted to think about the FDR among the 10 smallest:</p>
<p><span class="math display">\[\mathbb{E}_{F_X} \left [\frac{\# \{ \text{genes that are among 10 smallest, are rejected in }X\text{, and are null}\} }{\max \{1, \# \{\text{genes that are among 10 smallest and were rejected in } X \}} \right ]\]</span></p>
</div>
<div class="fragment">
<p>But the numerator and the denominator have both changed, making it impossible to argue that one of them is always smaller than the other!</p>
</div>
<div class="fragment">
<p>While the “biological pathway” version might feel intuitively more likely to be true, the logic above holds even when the subset of interest is fixed.</p>
<aside class="notes">
<ul>
<li>(the subset of interesting genes are not chosen based on the data)</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div>
</section>
<section id="subset-property-fwer" class="slide level1">
<h1>Subset property, FWER</h1>
<p>By contrast, <span class="math inline">\(\text{FWER} = \mathbb{P}_{F_X} [ \text{FD}(X) \geq 1 ]\)</span> <strong>does</strong> have the subset property.</p>
<div class="fragment">
<p>Consider:</p>
<ul>
<li>the number of false discoveries based on <span class="math inline">\(X\)</span> within all <span class="math inline">\(M\)</span> genes</li>
<li>the number of false discoveries based on <span class="math inline">\(X\)</span> within a subset of genes</li>
</ul>
<p>The former is ALWAYS larger!</p>
<p><br></p>
</div>
<div class="fragment">
<p>That means that if there is <span class="math inline">\(\geq 1\)</span> false discovery based on <span class="math inline">\(X\)</span> within a subset of genes, then there is <span class="math inline">\(\geq 1\)</span> false discovery based on <span class="math inline">\(X\)</span> within all <span class="math inline">\(M\)</span> genes. So</p>
<p><span class="math display">\[\text{FWER on subset} \leq \text{FWER on all}.\]</span></p>
</div>
<div class="fragment">
<p>This logic holds even if the gene subset is selected using the data; this has implications for selective inference (next two weeks!)</p>
</div>
</section>
<section id="do-we-need-to-adjust" class="slide level1">
<h1>Do we need to adjust?</h1>
<p>Let <span class="math inline">\(X \sim F_X\)</span> represent the data with which we test <span class="math inline">\(H_{0j}: T_j(F_X) = 0\)</span> for all <span class="math inline">\(j = 1, 2, \ldots, M\)</span>.</p>
<div class="fragment">
<p><br></p>
<p>The naive, unadjusted idea for screening here is:</p>
<ul>
<li>Test <span class="math inline">\(H_{0j}\)</span> for all <span class="math inline">\(j = 1, 2, \ldots, M\)</span>.</li>
<li>Reject all genes with p-value <span class="math inline">\(\leq \alpha\)</span>.</li>
</ul>
</div>
<div class="fragment">
<p><br></p>
<p>This will not control the FDR. We know this for sure because:</p>
<div>
<ul>
<li class="fragment">FWER = FDR when all tested hypotheses are null</li>
<li class="fragment">We established last week conditions where the naive approach does not control FWER.</li>
</ul>
</div>
</div>
<div class="fragment">
<p>(Not a very satisfying answer; we will get a more satisfying one on Thursday!)</p>
</div>
</section>
<section id="how-to-control-fdr" class="slide level1">
<h1>How to control FDR?</h1>
<p>We already know some ways, because FWER is always bigger than FDR. That means that any FWER controlling procedure controls FDR.</p>
<div class="fragment">
<p>But doing so would be unnecessarily conservative.</p>
</div>
<div class="fragment">
<p>We will focus on the landmark paper and method: Benjamini and Hochberg (1995).</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="fdr.png"></p>
</figure>
</div>
</div>
</div>
<div>
<ul>
<li class="fragment"><p>Showed that a procedure (independently developed by Eklund and Simes) controls the FDR.</p></li>
<li class="fragment"><p>This procedure is now commonly called the “Benjamini-Hochberg”, or BH procedure.</p></li>
</ul>
</div>
<aside class="notes">
<ul>
<li>One of the most cited papers in statistics!</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div>
</section>
<section id="bh-procedure" class="slide level1">
<h1>BH procedure</h1>
<p><strong>The idea</strong>: Order p-values <span class="math inline">\(p_{(1)} \leq \ldots \leq p_{(M)}\)</span>:</p>
<div>
<ul>
<li class="fragment"><p>If <span class="math inline">\(p_{(M)} \leq \alpha\)</span>, then reject <span class="math inline">\(H_{01}, \ldots, H_{0M}\)</span> and stop; otherwise keep going</p></li>
<li class="fragment"><p>If <span class="math inline">\(p_{(M-1)} \leq \alpha \left ( \frac{m-1}{m} \right )\)</span>, then reject <span class="math inline">\(H_{01}\)</span>, … <span class="math inline">\(H_{0(M-1)}\)</span> and stop; otherwise keep going</p></li>
<li class="fragment"><p>Keep going until you run out of p-values</p></li>
</ul>
</div>
<div class="fragment">
<p>Corresponds to the following rule: <span class="math display">\[\text{Reject}(p_1, \ldots, p_M) = \left \{ j: p_j \leq \max \{p_{(j)}: p_{(j)} \leq \frac{\alpha j}{m}  \right \}.\]</span></p>
</div>
</section>
<section id="comparing-rejection-boundaries" class="slide level1">
<h1>Comparing rejection boundaries</h1>

<img data-src="lec03-multiple-genomics_files/figure-revealjs/unnamed-chunk-4-1.png" width="960" class="r-stretch quarto-figure-center"><p>Here, Hochberg is a variant on Holm that controls FWER under independence; it uses the formula <span class="math inline">\(\text{Reject}(p_1, \ldots, p_M) = \left \{ j : p_j \leq \max \{p_{(j)}: p_{(j)} \leq \alpha/(M+1 - j)\} \right \}\)</span>.</p>
<aside class="notes">
<ul>
<li>Each method can be thought of as plotting the 100 p-values, finding the largest that falls below those lines, and rejecting everyting before that</li>
<li>At small indices <span class="math inline">\(j\)</span> (which we probably care about), BH is about <span class="math inline">\(j\)</span> times more liberal than Hochberg</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="bh-intuition" class="slide level1">
<h1>BH intuition</h1>
<p>Why does BH work? To get some intuition, we’ll set down a simple model and notation.</p>
<div class="fragment">
<p><br></p>
<p>Define <span class="math inline">\(H_{0j}: \mathcal{T}_j(F_X) = 0\)</span> for <span class="math inline">\(j = 1, 2, \ldots, M\)</span>, and let <span class="math inline">\(X \sim F_X\)</span> be the data that we use to test. Let <span class="math inline">\(M_0 = |\{j: F_X \in \mathcal{F}_{0j}\}|\)</span>.</p>
</div>
<div class="fragment">
<p><br></p>
<p>Let <span class="math inline">\(p_j(X) \sim \text{Uniform}(0, 1)\)</span> for all <span class="math inline">\(j\)</span> with <span class="math inline">\(H_{0j}\)</span> true, and <span class="math inline">\(p_j(X) \sim F_p\)</span> for all <span class="math inline">\(j\)</span> with <span class="math inline">\(H_{0j}\)</span> false.</p>
<p><br></p>
<p>We’ll use the notation <span class="math inline">\(\text{FD}(X; t)\)</span> to denote the number of false discoveries if you reject all p-values <span class="math inline">\(\leq t\)</span>.</p>
</div>
</section>
<section id="fdr-approximation" class="slide level1">
<h1>FDR approximation</h1>
<p>For any fixed threshold <span class="math inline">\(t\)</span> and <span class="math inline">\(M\)</span> big:</p>
<p><span class="math display">\[
\begin{align}
\text{FDR}_t &amp;= \mathbb{E}_{F_X} \left [\frac{\text{FD}(X; t)}{\max \{1, R(X; t)\}} \right ] \\
&amp;\approx \frac{\mathbb{E}_{F_X} [\text{FD}(X; t)]/M}{\mathbb{E}_{F_X} [R(X; t)]/M} \\
&amp;= \frac{\mathbb{E}_{F_X}  [ \sum_{j: H_{0j} \text{ true}} 1 \{p_j(X) \leq t \}   /m]}{\mathbb{E}_{F_X}  [ \sum_{j=1}^M 1 \{p_j(X) \leq t \}   /M]}\\
&amp;= \frac{M_0 \mathbb{P}(\text{Uniform}(0, 1) \leq t)/M}{\mathbb{E}_{F_X}  [ \sum_{j=1}^M 1 \{p_j(X) \leq t \}   /M]} = \frac{(M_0/M)t}{\mathbb{E}_{F_X}  [ \sum_{j=1}^M 1 \{p_j(X) \leq t \}   /M ]}
\end{align}
\]</span></p>
<aside class="notes">
<ul>
<li>First step: figure with <span class="math inline">\(M\)</span> big, we can forget about there being 0 discoveries</li>
<li>Approximate expectation of ratio with ratio of expectation</li>
<li>Seond step: definition</li>
<li>Third &amp; fourth: uniformity of p-values</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="estimating-fdr" class="slide level1">
<h1>Estimating FDR</h1>
<p><span class="math display">\[FDR_t \approx \frac{(M_0/M)t}{\mathbb{E}_{F_X}  [ \sum_{j=1}^M 1 \{p_j(X) \leq t \}  /M ]}\]</span></p>
<p>We don’t know what <span class="math inline">\(M_0/M\)</span> is, but we can conservatively estimate it by 1.</p>
<div class="fragment">
<p>We don’t know what <span class="math inline">\(\mathbb{E}_{F_X} [ \sum_{j=1}^M 1 \{p_j(X) \leq t \} /M ]\)</span> is, but given a realization <span class="math inline">\(x\)</span> from <span class="math inline">\(X \sim F_X\)</span>, we can estimate it with the empirical equivalent. Leads to:</p>
<p><span class="math display">\[\hat{FDR}_t(x) = \frac{t}{\sum_{j=1}^M 1 \{p_j(x) \leq t \}/M}.\]</span></p>
</div>
<div class="fragment">
<p>Assuming that there are no ties in p-values, and plugging in <span class="math inline">\(p_{(j)}(x)\)</span> to RHS yields <span class="math inline">\(\frac{p_{(j)}}{j/m}\)</span>.</p>
<aside class="notes">
<ul>
<li>generally speaking there won’t be ties with probabilty 1 if test statistic is continuous under the null and the alternative</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div>
</section>
<section id="reinterpretation-of-bh" class="slide level1">
<h1>Reinterpretation of BH</h1>
<p>Recall that BH says to reject every hypothesis with a p-value no larger than <span class="math inline">\(p_{(\hat j)}\)</span>, where<br>
<span class="math display">\[ \hat j = \max \left \{j: p_{(j)} \leq \frac{\alpha j}{M} \right \}.\]</span></p>
<div class="fragment">
<p>Rearranging terms, no larger than <span class="math display">\[ \hat j = \max \left \{j: \frac{p_{(j)}}{j/m} \leq \alpha \right \}\]</span></p>
</div>
<div class="fragment">
<p>Recall that <span class="math inline">\(\frac{p_{(j)}}{j/m}\)</span> is an estimate of the FDR if we reject at <span class="math inline">\(p_{(j)}\)</span>!</p>
<h3 id="bh-procedure-picks-the-biggest-rejection-threshold-that-yields-an-estimated-fdr-below-targer-alpha.">BH procedure picks the biggest rejection threshold that yields an estimated FDR below targer <span class="math inline">\(\alpha\)</span>.</h3>
</div>
</section>
<section id="empirical-bayes-interpretation" class="slide level1">
<h1>Empirical Bayes interpretation</h1>
<p>Do you like Bayesian measures of uncertainty? Augment our simple model with a prior:</p>
<p><span class="math display">\[ 1\{ H_{0j} \text{ false}\} \overset{iid}{\sim} \text{Bernoulli}(\pi_0)\]</span></p>
<div class="fragment">
<p>Then for any threshold <span class="math inline">\(t\)</span>, can derive the “Bayesian FDR”:<br>
<span class="math display">\[\mathbb{P}(H_{0j} \text{ true} \mid p_j(x) \leq t) = \frac{t\pi_0}{\mathbb{P}(p_j(X) \leq t)}. \]</span> An empirical Bayesian then estimates with frequentist principles to get: <span class="math display">\[\mathbb{P}(H_{0j} \text{ true} \mid p_j(x) \leq t) \approx \frac{t}{\sum_{j'=1}^M 1 \{p_{j'}(x) \leq t \}/M}. \]</span></p>
<aside class="notes">
<ul>
<li>Just Bayes rule</li>
<li>“If we report a null as false, then what’s the chance that it’s a false discovery?”</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div>
</section>
<section id="empirical-bayes-interpretation-1" class="slide level1">
<h1>Empirical Bayes interpretation</h1>
<p><span class="math display">\[\mathbb{P}(H_{0j} \text{ true} \mid p_j(x) \leq t) \approx \frac{t}{\sum_{j=1}^M 1 \{p_j(x) \leq t \}/M}. \]</span> This is our frequentist estimate of the approximation to <span class="math inline">\(FDR_t\)</span> from earlier!</p>
<div class="fragment">
<p><br></p>
<p>So another interpretation is that BH takes the largest threshold so that the estimated Bayesian FDR is <span class="math inline">\(\leq \alpha\)</span>.</p>
</div>
<div class="fragment">
<p><br></p>
<p>If we buy into Bayesian notions of probability, can even say rejected cases “have estimated probability of being null lower than <span class="math inline">\(\alpha\)</span>”.</p>
</div>
</section>
<section id="bh-under-independence" class="slide level1">
<h1>BH under independence</h1>
<p>Benjamini and Hochberg (1995) showed the following result.</p>
<p><br></p>
<p>Suppose that we test <span class="math inline">\(H_{0j}\)</span> for <span class="math inline">\(j = 1, 2, \ldots, M\)</span> using <span class="math inline">\(X \sim F_X\)</span>, and choose which hypotheses to reject by plugging p-values <span class="math inline">\(p_1(X), \ldots, p_M(X)\)</span> into the BH procedure.</p>
<div class="fragment">
<p><br> Suppose that <span class="math inline">\(p_1(X), \ldots, p_M(X)\)</span> are valid .</p>
</div>
<div class="fragment">
<p><br></p>
<p>Then, <span class="math display">\[\text{FDR}_{BH} = \frac{M_0}{M} \alpha \leq \alpha.\]</span></p>
</div>
</section>
<section id="bh-under-prds" class="slide level1">
<h1>BH under PRDS</h1>
<p>Benjamini and Yekeutili (2001) showed that:</p>
<p><br></p>
<p>If the joint distribution of the p-values is independent, OR the subset of p-values corresponding to true nulls is ``positive regression dependence on a subset” (PRDS), then <span class="math display">\[ \text{FDR}_{BH} = \frac{M_0}{M} \alpha \leq \alpha.\]</span></p>
<div class="fragment">
<p>PRDS is technical and not so interpretable, but can show it holds for important subcases: e.g.&nbsp;p-values constructed with test statistics where the null distributions are jointly multivariate normal with all off-diagonal covariance matrix entries positive.</p>
</div>
</section>
<section id="bh-under-general-dependence" class="slide level1">
<h1>BH under general dependence</h1>
<div>
<ul>
<li class="fragment">Benjamini and Yekeutili (2001) also showed it’s possible to make a small change to BH procedure to control the FDR under ANY dependence structure</li>
<li class="fragment"><code>p.adjust(..., method = "BY")</code> in R</li>
<li class="fragment">But this is typically very very conservative; not often used</li>
<li class="fragment">In practice, BH is hard to “break”: FDR is typically controlled even if PRDS is not satisfied</li>
<li class="fragment">That being said, correlation can still throw a wrench by increasing variability of FDP. (Might see some of this on Thursday … )</li>
</ul>
</div>
</section>
<section id="adjusted-p-values" class="slide level1">
<h1>Adjusted p-values</h1>
<p><strong>The idea</strong>: Convert <span class="math inline">\(p_1, \ldots, p_M\)</span> to <span class="math inline">\(\tilde p_1, \ldots, \tilde p_M\)</span> so that <span class="math display">\[\text{Reject}(p_1, \ldots, p_M) = \{j: \tilde p_j \leq \alpha\}\]</span></p>
<p>See e.g.&nbsp;<code>stats::p.adjust()</code>. General definition that applies to any rejection procedure.</p>
<div class="fragment">
<h3 id="caution-interpretation-issues">Caution: interpretation issues</h3>
<div>
<ul>
<li class="fragment"><p>Raw p-values have a natural interpretation: “If the null hypothesis were true, then how often would I see such an extreme value of the test statistic, were I to <em>repeat</em> the experiment?”.</p></li>
<li class="fragment"><p>Justifies interpretation as “measure of evidence against the null”, and justifies thumb rules (&lt; 0.001 = “overwhelming”, &lt; 0.01 = “strong”, &lt; 0.05 = “moderate”)</p></li>
<li class="fragment"><p>Adjusted p-values have no such interpretation; so what can we read into (say) <span class="math inline">\(\tilde p_j = 0.035\)</span>?</p></li>
<li class="fragment"><p>In the case of FDR adjustment methods, q-values provide an interpretable alternative; see <a href="https://www.pnas.org/doi/10.1073/pnas.1530509100">Storey and Tibshirani (2003)</a> if you’re interested in learning more.</p></li>
</ul>
</div>
<p><br></p>
<aside class="notes">
<ul>
<li>It currently seems to me that the only role adjusted p-values have to play is making calculation easier for the practicioner - seems not so relevant?</li>
<li>(They are apparently also convenienet to work with in methods development)</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div>
</section>
<section id="food-for-thought" class="slide level1">
<h1>Food for thought</h1>
<div>
<ul>
<li class="fragment">We’ve really only scratched the surface; this was and is a very active research area!</li>
<li class="fragment">There are many other proposed error rates: positive FDR, directional FDR</li>
<li class="fragment">There are related research directions: e.g.&nbsp;exceedance probability control</li>
<li class="fragment">There are many methods for controlling FDR that try to improve on BH, e.g.&nbsp;by estimating the number of truly null hypotheses from the data</li>
<li class="fragment">There are many methods for controlling FDR that look quite different; a particularly popular one in recent years is knockoffs</li>
</ul>
</div>
<div class="fragment">
<h3 id="many-options-you-can-pick-for-your-final-presentation-that-can-help-you-learn-more">Many options you can pick for your final presentation that can help you learn more!</h3>
</div>
</section>
<section id="multiple-testing-wrapup" class="slide level1">
<h1>Multiple testing wrapup</h1>
<p><strong>Multiple testing</strong>: Whenever we conduct more than one hypothesis test in a single analysis, and look at the outcome of all of them.</p>
<ul>
<li>Does this cause problems? If so, what are they?</li>
<li>Do we need to adjust for multiple testing? If so, how should we adjust?</li>
</ul>
<p><br></p>
<div class="fragment">
<p>We saw very different answers to these questions within a single class last week, and then once again very different answers this week!</p>
<p><br></p>
</div>
<div class="fragment">
<p>Does that make sense?</p>
</div>
</section>
<section id="a-review-of-the-science" class="slide level1">
<h1>A review of the science</h1>
<blockquote>
<p><strong>“I don’t know, it depends on the scientific problem at hand.”</strong> - My annoying partner, whenever I ask him ANY question involving statistics.</p>
</blockquote>
<table>
<thead>
<tr class="header">
<th></th>
<th>Clinical Trials</th>
<th>Genomics</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Number of tests?</td>
<td>A few</td>
<td>Lots</td>
</tr>
<tr class="even">
<td>Hypothesis selection</td>
<td>Carefully curated</td>
<td>Little to no screening</td>
</tr>
<tr class="odd">
<td>False positives?</td>
<td>Awful</td>
<td>Can live with a few</td>
</tr>
</tbody>
</table>
<div class="fragment">
<p><br></p>
<p>The science is different, so no wonder the answers are different.</p>
<blockquote>
<p><strong>“You better think (think) about what you’re trying to do …”</strong> - Aretha Franklin, “Think”</p>
</blockquote>
<aside class="notes">
<ul>
<li>One hypothesis per measured endpoint/outcome.</li>
<li>Last stage before the treatment goes on market</li>
<li>Results inform market labelling</li>
<li>Treatments go through a very extensive screening process before reaching Phase III</li>
<li>Endpoints are very carefully chosen based on clinical considerations</li>
<li>Federal agency involving medical considerations: ENORMOUS potential for political fallout</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
<div class="footer footer-default">

</div>
</div>
</section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="lec03-multiple-genomics_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="lec03-multiple-genomics_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="lec03-multiple-genomics_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="lec03-multiple-genomics_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="lec03-multiple-genomics_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="lec03-multiple-genomics_files/libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="lec03-multiple-genomics_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="lec03-multiple-genomics_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="lec03-multiple-genomics_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="lec03-multiple-genomics_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="lec03-multiple-genomics_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1600,

        height: 900,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>