<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Lab 3: RNA Sequencing Data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="lab03_files/libs/clipboard/clipboard.min.js"></script>
<script src="lab03_files/libs/quarto-html/quarto.js"></script>
<script src="lab03_files/libs/quarto-html/popper.min.js"></script>
<script src="lab03_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="lab03_files/libs/quarto-html/anchor.min.js"></script>
<link href="lab03_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="lab03_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="lab03_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="lab03_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="lab03_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#scientific-context" id="toc-scientific-context" class="nav-link active" data-scroll-target="#scientific-context">Scientific context</a></li>
  <li><a href="#task-1" id="toc-task-1" class="nav-link" data-scroll-target="#task-1">Task 1</a></li>
  <li><a href="#task-2" id="toc-task-2" class="nav-link" data-scroll-target="#task-2">Task 2</a></li>
  <li><a href="#code" id="toc-code" class="nav-link" data-scroll-target="#code">Code</a>
  <ul class="collapse">
  <li><a href="#data-generation" id="toc-data-generation" class="nav-link" data-scroll-target="#data-generation">Data generation</a></li>
  <li><a href="#data-analysis" id="toc-data-analysis" class="nav-link" data-scroll-target="#data-analysis">Data analysis</a></li>
  </ul></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Lab 3: RNA Sequencing Data</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p><strong>Grading</strong>: Turn in your first attempt at the tasks for a binary “fair attempt or not” grade on Canvas. That is, your first attempt need not be neat or correct.</p>
<p><strong>Done early?</strong> Come discuss your work with me and I’ll give you one suggestion on how to improve what you’ve got in the remaining time. Iterate if necessary.</p>
<section id="scientific-context" class="level2">
<h2 class="anchored" data-anchor-id="scientific-context">Scientific context</h2>
<p>Consider the following RNA sequencing experiment conducted by your collaborator:</p>
<ul>
<li><p><em>Data:</em> A count matrix with one row for each (biological) sample, one column for each gene, and <span class="math inline">\((i, j)\)</span>th entry indicating the number of sequencing reads that are attributed to gene <span class="math inline">\(j\)</span> in sample <span class="math inline">\(i\)</span>; and a condition label for each sample (A or B).</p></li>
<li><p><em>Analysis:</em> For each gene <span class="math inline">\(j\)</span>, test for (multiplicative) differential expression by fitting a Poisson GLM with condition as the covariate, and testing the null hypothesis that the condition coefficient is equal to 0.</p></li>
</ul>
<p>(Note that the data model and the analysis today are heavily simplified and are not exactly what you will see out in the wild!)</p>
<p>Your collaborator wants to call any gene that has p-value smaller than 0.1 differentially expressed. Their ultimate concern is the proportion of follow-up validation efforts from their RNA-seq experiment that is ultimately wasted. To help them better understand the consequences of this strategy, you decide to design a simulation study to investigate the distribution of the false discovery proportion.</p>
<p>Functions to generate and analyze a single RNA-seq data set on 100 samples (50 from each condition) and 100 genes are provided at the end of this document. You are free to modify them if you see fit.</p>
</section>
<section id="task-1" class="level2">
<h2 class="anchored" data-anchor-id="task-1">Task 1</h2>
<p>Decide on a single set of values for the arguments of <code>generate_RNAseq_data()</code>; think of this like fixing a particular RNA-seq experiment to repeat. Simulate differential expression results from 200 repetitions of this experiment, and summarize the FDP distribution across the 200 repetitions.</p>
</section>
<section id="task-2" class="level2">
<h2 class="anchored" data-anchor-id="task-2">Task 2</h2>
<p>Simulate to investigate how (i) the proportion of non-differentially expressed genes and (ii) the correlation between genes affect the FDP distribution. Summarize your results with a plot.</p>
</section>
<section id="code" class="level2">
<h2 class="anchored" data-anchor-id="code">Code</h2>
<section id="data-generation" class="level3">
<h3 class="anchored" data-anchor-id="data-generation">Data generation</h3>
<p>Arguments:</p>
<ul>
<li><code>prop_null</code>: the proportion of non-differentially expressed genes; for e.g.&nbsp;<code>prop_null = 90</code>, there will be <span class="math inline">\(0.9*100 = 90\)</span> differentially expressed genes.</li>
<li><code>de_effect</code>: degree of differential expression; for e.g.&nbsp;<code>de_effect = 2</code>, the differentially expressed genes have a mean expression level twice as big in condition B compared to condition A</li>
<li><code>rho</code>: amount of dependence; for e.g.&nbsp;<code>rho = 0.5</code>, the correlation between two genes in each condition is 0.5.</li>
</ul>
<p>I’m labelling this as “model and data” because I’m separately saving out some information about the generating model that we would not know in real life (which genes are differentially expressed).</p>
<p>In practice, I usually set up my simulations with the pipeline <code>generate_model() %&gt;% generate_data_from_model() %&gt;% analyze_data() %&gt;% evaluate_results()</code>. (In part because I’m used to using the <a href="https://faculty.marshall.usc.edu/jacob-bien/simulator.html">simulator</a> package, and in part because I just like the flexibility of designing each component of the infrastructure of a simulation study separately.)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>generate_model_and_data <span class="ot">&lt;-</span> <span class="cf">function</span>(prop_null, de_effect, rho) { </span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>    n <span class="ot">&lt;-</span> <span class="dv">50</span> </span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>    p <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>    num_null <span class="ot">&lt;-</span> <span class="fu">ceiling</span>(prop_null<span class="sc">*</span>p)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>    cond <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"A"</span>, n), <span class="fu">rep</span>(<span class="st">"B"</span>, n)))</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>    Sigma <span class="ot">&lt;-</span> (<span class="dv">1</span><span class="sc">-</span>rho)<span class="sc">*</span><span class="fu">diag</span>(p) <span class="sc">+</span> rho<span class="sc">*</span><span class="fu">matrix</span>(<span class="dv">1</span>, p, p)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>    mvn_copula_gen <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(<span class="dv">2</span><span class="sc">*</span>n<span class="sc">*</span>p), <span class="dv">2</span><span class="sc">*</span>n, p, <span class="at">byrow=</span><span class="cn">TRUE</span>)<span class="sc">%*%</span><span class="fu">chol</span>(Sigma) <span class="sc">%&gt;%</span> </span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>                                            <span class="fu">as_tibble</span>(<span class="at">.name_repair =</span> \(x) <span class="fu">str_c</span>(<span class="st">"gene"</span>, <span class="dv">1</span><span class="sc">:</span>p)) <span class="sc">%&gt;%</span> </span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>                                            <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">everything</span>(), \(x) <span class="fu">pnorm</span>(x)))</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>    pois_data_gen <span class="ot">&lt;-</span> mvn_copula_gen <span class="sc">%&gt;%</span> </span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>            <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">everything</span>(),         </span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>                                <span class="cf">function</span>(x) { </span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>                                        <span class="fu">ifelse</span>(cond <span class="sc">==</span> <span class="st">"B"</span> <span class="sc">&amp;</span> </span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>                                                     <span class="fu">as.numeric</span>(<span class="fu">str_remove</span>(<span class="fu">cur_column</span>(), <span class="st">"gene"</span>))   <span class="sc">&gt;</span> num_null, </span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>                                                     <span class="fu">qpois</span>(x, <span class="at">lambda=</span><span class="dv">5</span><span class="sc">*</span>de_effect), </span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>                                                     <span class="fu">qpois</span>(x, <span class="at">lambda=</span><span class="dv">5</span>))</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>                                })) <span class="sc">%&gt;%</span> </span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a>            <span class="fu">mutate</span>(<span class="at">cond=</span>cond) <span class="sc">%&gt;%</span> </span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>            <span class="fu">select</span>(cond, <span class="fu">everything</span>())</span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a>  gene <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"H0"</span>, num_null), <span class="fu">rep</span>(<span class="st">"H1"</span>, p <span class="sc">-</span> num_null))</span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(<span class="at">data=</span>pois_data_gen,<span class="at">gene=</span>gene))</span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2</span>) </span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a>ex_model_and_data <span class="ot">&lt;-</span> <span class="fu">generate_model_and_data</span>(<span class="at">prop_null=</span><span class="fl">0.9</span>, <span class="at">de_effect=</span><span class="dv">2</span>, <span class="at">rho=</span><span class="dv">0</span>)</span>
<span id="cb1-36"><a href="#cb1-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-37"><a href="#cb1-37" aria-hidden="true" tabindex="-1"></a>ex_model_and_data<span class="sc">$</span>data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 100 × 101
   cond  gene1 gene2 gene3 gene4 gene5 gene6 gene7 gene8 gene9 gene10 gene11
   &lt;fct&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
 1 A         3     5     9     3     5     5     6     4    10      5      6
 2 A         7     5     4     3     3    10     7    10     4      4      3
 3 A         6     3    12     5     3     6     5     4     6      5      7
 4 A         4     4     7     1     7     7     5     7     8      3      6
 5 A         4     0     3     3     7     4     3     1     8      5      5
 6 A         4     6     3     4     5     8     3     8     3      8      2
 7 A         3     5     3     3     8     7     5     9     7      3      4
 8 A         8     5     4     9     5     8     5     5     3      5      1
 9 A         3     1     6     4     5     3     7    11     8      4      4
10 A         5     7     5     5     5     6     6     5     3      4     10
# ℹ 90 more rows
# ℹ 89 more variables: gene12 &lt;dbl&gt;, gene13 &lt;dbl&gt;, gene14 &lt;dbl&gt;, gene15 &lt;dbl&gt;,
#   gene16 &lt;dbl&gt;, gene17 &lt;dbl&gt;, gene18 &lt;dbl&gt;, gene19 &lt;dbl&gt;, gene20 &lt;dbl&gt;,
#   gene21 &lt;dbl&gt;, gene22 &lt;dbl&gt;, gene23 &lt;dbl&gt;, gene24 &lt;dbl&gt;, gene25 &lt;dbl&gt;,
#   gene26 &lt;dbl&gt;, gene27 &lt;dbl&gt;, gene28 &lt;dbl&gt;, gene29 &lt;dbl&gt;, gene30 &lt;dbl&gt;,
#   gene31 &lt;dbl&gt;, gene32 &lt;dbl&gt;, gene33 &lt;dbl&gt;, gene34 &lt;dbl&gt;, gene35 &lt;dbl&gt;,
#   gene36 &lt;dbl&gt;, gene37 &lt;dbl&gt;, gene38 &lt;dbl&gt;, gene39 &lt;dbl&gt;, gene40 &lt;dbl&gt;, …</code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(ex_model_and_data<span class="sc">$</span>gene)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
H0 H1 
90 10 </code></pre>
</div>
</div>
</section>
<section id="data-analysis" class="level3">
<h3 class="anchored" data-anchor-id="data-analysis">Data analysis</h3>
<p>The following function takes in a RNA-sequencing data set <code>data</code> of the format produced by <code>generate_model_and_data()</code> and returns a p-value for each gene by fitting a Poisson GLM with condition as the covariate, and testing the null hypothesis that the condition coefficient is equal to 0, and a rejection status for each gene based on thresholding the p-value at <code>alpha</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>analyze_RNAseq_data <span class="ot">&lt;-</span> <span class="cf">function</span>(data, alpha) { </span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>        data <span class="sc">%&gt;%</span> </span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>                <span class="fu">summarize</span>(<span class="fu">across</span>(<span class="sc">-</span>cond, <span class="cf">function</span>(gene) broom<span class="sc">::</span><span class="fu">tidy</span>(<span class="fu">glm</span>(gene <span class="sc">~</span> cond), </span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>                                <span class="at">family=</span><span class="st">"poisson"</span>)[[<span class="st">"p.value"</span>]][<span class="dv">2</span>])) <span class="sc">%&gt;%</span> </span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>                                    tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">everything</span>(), <span class="at">names_to =</span> <span class="st">"gene"</span>, </span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>                                                                            <span class="at">values_to =</span> <span class="st">"pval"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>                                    <span class="fu">mutate</span>(<span class="at">rejects =</span> (pval <span class="sc">&lt;=</span> alpha))</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="fu">analyze_RNAseq_data</span>(ex_model_and_data<span class="sc">$</span>data, <span class="fl">0.1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 100 × 3
   gene     pval rejects
   &lt;chr&gt;   &lt;dbl&gt; &lt;lgl&gt;  
 1 gene1  0.293  FALSE  
 2 gene2  0.963  FALSE  
 3 gene3  0.0462 TRUE   
 4 gene4  0.0646 TRUE   
 5 gene5  0.314  FALSE  
 6 gene6  0.245  FALSE  
 7 gene7  0.239  FALSE  
 8 gene8  0.541  FALSE  
 9 gene9  1.00   FALSE  
10 gene10 0.917  FALSE  
# ℹ 90 more rows</code></pre>
</div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>